pkg/controller/deployment/deployment_controller.go

```diff
@@ -576,6 +579,73 @@ func (dc *DeploymentController) syncRollingUpdateDeployment(deployment extension
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// proportionalScaling should run only on scaling events and not during the normal rollout process.
+func (dc *DeploymentController) proportionalScaling(deployment extensions.Deployment, shouldCreate bool) error {
+	newRS, oldRSs, err := dc.getAllReplicaSets(deployment, shouldCreate)
+	if err != nil {
+		return err
+	}
+	allRSs := controller.FilterActiveReplicaSets(oldRSs)
+	maybeMaxSurge := 0
+	if deployment.Spec.Strategy.Type == extensions.RollingUpdateDeploymentStrategyType {
+		maybeMaxSurge, err = intstrutil.GetValueFromIntOrPercent(&deployment.Spec.Strategy.RollingUpdate.MaxSurge, deployment.Spec.Replicas)
+		if err != nil {
+			return err
+		}
+	}
+
+	if newRS != nil {
+		// No old replica sets with pods; scale new replica set to the full count.
+		if len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+			if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+				return err
+			}
+			deployment.Status.EnactedReplicas = deployment.Spec.Replicas + maybeMaxSurge
+			return dc.updateDeploymentStatus(allRSs, newRS, deployment)
+		}
+		allRSs = append(allRSs, newRS)
+		// Normalize maxSurge in case newRS replicas exceed deployment replicas.
+		maybeMaxSurge = integer.IntMax(integer.IntMin(maybeMaxSurge, deployment.Spec.Replicas-newRS.Spec.Replicas), 0)
+	}
+
+	sort.Sort(controller.ByAvailableReplicas(allRSs))
+
+	totalReplicas := deploymentutil.GetReplicaCountForReplicaSets(allRSs)
+	totalAdded := 0
+	// Number of additional pods that can be either added or removed from the total pod count.
+	additional := deployment.Spec.Replicas - totalReplicas
```

Comment:
IIUC, surge is meant to allow rollouts to proceed but eventually the replica size should boil down to spec.replicas, right? Especially here we just want to scale to a desired amount, I don't see any value in surge but maybe I am wrong.